### PR TITLE
{bio}[foss/2023b] GROMACS v2024.3

### PR DESCRIPTION
**Pull request by bedroge**
(created using `eb --new-pr`)

Compared to previous easyconfigs, this now installs the pypi version of gmxapi. The versioning of the included gmxapi seems a bit confusing: https://gitlab.com/gromacs/gromacs/-/blob/v2024.3/python_packaging/gmxapi/pyproject.toml?ref_type=tags says 0.4.1, https://gitlab.com/gromacs/gromacs/-/blob/v2024.3/python_packaging/gmxapi/src/gmxapi/version.py?ref_type=tags shows 0.5.0a1, and the docs just recommend using the pypi version (where the latest version is 0.4.2).

### CODE REVIEW

**Comment by bedroge** [comment]
@boegelbot please test @ jsc-zen3
CORE_CNT=16

**Comment by bedroge** [comment]
Test report by @bedroge
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
bob-Latitude-5300 - Linux Ubuntu 24.04.1 LTS (Noble Numbat), x86_64, Intel(R) Core(TM) i7-8665U CPU @ 1.90GHz, Python 3.12.3
See https://gist.github.com/bedroge/00b77ed6bd3a5d428ec87908696c72e3 for a full test report.

**Comment by bedroge** [comment]
@boegelbot please test @ generoso
CORE_CNT=16

**Comment by bedroge** [comment]
@boegelbot please test @ generoso
CORE_CNT=16

**Comment by bedroge** [comment]
Test report by @bedroge
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/3283
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
bob-Latitude-5300 - Linux Ubuntu 24.04.1 LTS (Noble Numbat), x86_64, Intel(R) Core(TM) i7-8665U CPU @ 1.90GHz, Python 3.12.3
See https://gist.github.com/bedroge/3ef052bc3b70ea4b0c73ab4eb450ade9 for a full test report.

**Comment by bedroge** [comment]
Also tested this with the EESSI bot for a bunch of CPUs: https://github.com/EESSI/software-layer/pull/709. There it also failed on haswell with the same input/output error, so I've started another build.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
node3105.skitty.os - Linux RHEL 8.8, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 3.6.8
See https://gist.github.com/boegel/1a1c9f63138cf054dbb09e6d2b83ea0a for a full test report.

**Comment by mabraham** [comment]
> Test report by @boegel **FAILED** Build succeeded for 0 out of 1 (1 easyconfigs in total) node3105.skitty.os - Linux RHEL 8.8, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 3.6.8 See https://gist.github.com/boegel/1a1c9f63138cf054dbb09e6d2b83ea0a for a full test report.

GROMACS dev here. I see that the following test case

```
[  FAILED  ] PropagatorsWithCoupling/PeriodicActionsTest.PeriodicActionsAgreeWithReference/21, where GetParam() = ({ ("comm-mode", "linear"), ("integrator", "md-vv"), ("maxGromppWarningsTolerated", "0"), ("nstcomm", "5"), ("nstpcouple", "3"), ("nsttcouple", "2"), ("pcoupl", "C-rescale"), ("simulationName", "argon12"), ("tcoupl", "v-rescale") }, 0x55e03938e489)
```

fails, either timing out or somehow suspended or crashed. C-rescale is a relatively new implementation, and this test case is intended to exercise dark corners of the code, so a real problem is possible.

Yet I see the preceding test case (at https://gist.github.com/boegel/75ff6503735f73f2d9ec570366bd181f#file-gromacs-2024-3-foss-2023b_partial-log-L374) took 25 seconds. On my x86 laptop with a release build the whole test suite takes under 4 seconds. Why is this GROMACS configuration so slow?

**Comment by boegel** [comment]
> > Test report by @boegel **FAILED** Build succeeded for 0 out of 1 (1 easyconfigs in total) node3105.skitty.os - Linux RHEL 8.8, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 3.6.8 See https://gist.github.com/boegel/1a1c9f63138cf054dbb09e6d2b83ea0a for a full test report.
> 
> GROMACS dev here. I see that the following test case
> 
> ```
> [  FAILED  ] PropagatorsWithCoupling/PeriodicActionsTest.PeriodicActionsAgreeWithReference/21, where GetParam() = ({ ("comm-mode", "linear"), ("integrator", "md-vv"), ("maxGromppWarningsTolerated", "0"), ("nstcomm", "5"), ("nstpcouple", "3"), ("nsttcouple", "2"), ("pcoupl", "C-rescale"), ("simulationName", "argon12"), ("tcoupl", "v-rescale") }, 0x55e03938e489)
> ```
> 
> fails, either timing out or somehow suspended or crashed. C-rescale is a relatively new implementation, and this test case is intended to exercise dark corners of the code, so a real problem is possible.
> 
> Yet I see the preceding test case (at https://gist.github.com/boegel/75ff6503735f73f2d9ec570366bd181f#file-gromacs-2024-3-foss-2023b_partial-log-L374) took 25 seconds. On my x86 laptop with a ~release~ debug build the whole test suite takes under 4 seconds. Why is this GROMACS configuration so slow?

@mabraham It probably not the GROMACS configuration itself, but the environment it's running it.

It's running in an interactive Slurm job, with 9 cores available (in a cgroup) out of a total of 36 in total on that system.
It's also an Intel Skylake system (Intel Xeon Gold 6140), which isn't exactly new.

In addition, `$OMP_PROC_BIND` is set to `TRUE` by default on that system (via a profile script).
In general, that should improve performance for OpenMP workloads, but we've seen that cause trouble before: for some software multi-threaded processes are being bound to a single core (that's definitely the case for R example), while they do start N threads, so threads are fighting for resources leading to very slow runs.
That's a known quirk of the GCC OpenMP runtime, see https://gcc.gnu.org/bugzilla/show_bug.cgi?id=113698

I've seen this before, but I never got to the bottom of it for GROMACS...

If any of this rings a bell, any insights you may have are welcome.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
node3108.skitty.os - Linux RHEL 9.4, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz (skylake_avx512), Python 3.9.18
See https://gist.github.com/boegel/541a0814c04255d0d7c33dd085ae4f9c for a full test report.

**Comment by boegel** [comment]
The system I was testing on has been migrated from RHEL 8.8 to RHEL 9.4 since the last time I tested (17 Sept'24), and the last attempt didn't fail (see test report above)

One difference there is that this test was done on a full workernode (all 36 cores assigned to the Slurm job), so there's no cgroup effect here. I also did an `unset OMP_PROC_BIND` before starting the test.

I'm now retesting in a 9-core Slurm job, where cgroup is set up such that available cores are spread across the node:
```
$ taskset -c -p $$
pid 3455741's current affinity list: 3,7,11-13,15,19,23,33
```

I didn't see any failing tests after an `unset OMP_PROC_BIND` in that environment either.

If I keep `$OMP_PROC_BIND` set to `TRUE` however, I do this:
```
      Start 80: MdrunCoordinationCouplingTests2Ranks
80/89 Test #80: MdrunCoordinationCouplingTests2Ranks .........***Timeout 480.00 sec
...
The following tests FAILED:
         80 - MdrunCoordinationCouplingTests2Ranks (Timeout)
```

Seems to be the same `PropagatorsWithCoupling/PeriodicActionsTest.PeriodicActionsAgreeWithReference/21` as before.

That's not a total surprise though, we've seen other situations where setting `$OMP_PROC_BIND` to `TRUE` means trouble...

So long story short: friends don't let friends set `$OMP_PROC_BIND` to `TRUE` when running the GROMACS test suite?

@mabraham Not sure if it makes sense to integrate the "unset `OMP_PROC_BIND` if set" in the GROMACS test suite or not...

**Comment by mabraham** [comment]
It certainly makes sense for you to integrate that unset call in your runner. By default, GROMACS does try to respect existing thread-affinity settings, but if it detects none, then it sets them itself. The main simulation engine itself has a command-line flag to specify behavior here, but the these test binaries just do the default. However the default only checks GOMP_CPU_AFFINITY, and no OMP_* variables, which looks like an omission.

**Comment by mabraham** [comment]
I made https://gitlab.com/gromacs/gromacs/-/issues/5170 to follow up

File: easybuild/easyconfigs/g/GROMACS/GROMACS-2024.3-foss-2023b.eb [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+exts_default_options = {
+    'source_urls': [PYPI_SOURCE],
+    'use_pip': True,
+    'download_dep_fail': True,
+    'sanity_pip_check': True,
+}
+
+exts_list = [
+    ('gmxapi', '0.4.2', {
```
Comment (boegel) [inline easybuild/easyconfigs/g/GROMACS/GROMACS-2024.3-foss-2023b.eb:74]: @bedroge This is using an older version of `gmxapi` compared to `GROMACS-2024.1-foss-2023b.eb`, see also https://github.com/easybuilders/easybuild-easyconfigs/pull/20522

Also, we're getting the `gmxapi` from PyPI here, while with GROMACS 2024.1 we used the one included with GROMACS -- we should probably keep doing that?

**Review by boegel — CHANGES_REQUESTED** [review]

File: easybuild/easyconfigs/g/GROMACS/GROMACS-2024.3-foss-2023b.eb (hunk 1 above)
Comment (bedroge) [reply easybuild/easyconfigs/g/GROMACS/GROMACS-2024.3-foss-2023b.eb:74]: I motivated that change in the PR description (see https://github.com/easybuilders/easybuild-easyconfigs/pull/21430#issue-2530991624). 

Another weird example/behavior of the old approach can be seen in the following inconsistencies (based on the 2024.1 version in EESSI):
```
$ ml GROMACS/2024.1-foss-2023b

$ ls $EBROOTGROMACS/lib/python3.11/site-packages/
gmxapi  gmxapi-0.4.1.dist-info
$ cat $EBROOTGROMACS/lib/python3.11/site-packages/gmxapi-0.4.1.dist-info/METADATA  | grep "^Version"
Version: 0.4.1
$ python -c "from importlib.metadata import version; print(version('gmxapi'))"
0.4.1

$ python -c "import gmxapi; print(gmxapi.__version__)"
0.5.0a1
$ python -c "import gmxapi; print(gmxapi.version.__version__)"
0.5.0a1
```

Comment (boegel) [reply easybuild/easyconfigs/g/GROMACS/GROMACS-2024.3-foss-2023b.eb:74]: OK, since the installation instructions also recommend installing the latest `gmxapi` version from PyPI (see [here](https://manual.gromacs.org/current/gmxapi/userguide/install.html#install-the-gmxapi-python-package)), we should indeed use this approach.

Thanks for clarifying!

**Review by boegel — APPROVED** [review]
lgtm

**Comment by boegel** [comment]
Going in, thanks @bedroge!
